src/main/java/io/confluent/idesidecar/restapi/cache/Clients.java

```diff
@@ -0,0 +1,129 @@
+package io.confluent.idesidecar.restapi.cache;
+
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.github.benmanes.caffeine.cache.CaffeineSpec;
+import com.github.benmanes.caffeine.cache.RemovalListener;
+import io.confluent.idesidecar.restapi.connections.ConnectionState;
+import io.confluent.idesidecar.restapi.events.Lifecycle;
+import io.quarkus.logging.Log;
+import jakarta.enterprise.event.ObservesAsync;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Supplier;
+
+/**
+ * Utilities to obtain and cache clients for a given connection and client ID.
+ */
+public abstract class Clients<T extends AutoCloseable> {
+
+  /**
+   * Store an instance of a Caffeine cache for each connection. The cache will store
+   * clients by client ID and its policy may be configured by the {@link #getCaffeineSpec()}
+   * method.
+   */
+  private final Map<String, Cache<String, T>> clientsByIdByConnections =
+      new ConcurrentHashMap<>();
+
+  RemovalListener<String, T> removalListener = (key, value, cause) -> {
+    try {
+      if (value != null) {
+        Log.debugf("Closing client %s", value);
+        value.close();
+      }
```

Comment:
What do you think about making this function a member method, like:
```
   void handleRemoval(String key, T value, RemovalCause cause) { ... }
```
Then `createCache()` would reference the method `this::handleRemoval`. (This should work since `createCache()` is not called during the `AdminClient` constructor.)